src/Materials/Textures/Packer/packer.ts

```diff
@@ -0,0 +1,792 @@
+                            }
+                            continue;
+                        }
+
+                        material.forceCompilationAsync(mesh).then(() => {
+                            doneCheck((material as Material));
+                        });
+                    }
+                }catch (e) {
+                    return reject(e);
+                }
+            });
+    }
+
+    /**
+    * Updates a Meshes materials to use the texture packer channels
+    * @param m is the mesh to target
+    */
+    public dispose(): void {
+        let sKeys = Object.keys(this.sets);
+        for (let i = 0; i < sKeys.length; i++) {
+            let channel = sKeys[i];
+            (this.sets as any)[channel].dispose();
+        }
+    }
+
+    /**
+    * Starts the download process for all the channels converting them to base64 data and embedding it all in a JSON file.
+    * @param imageType is the image type to use.
+    * @param quality of the image if downloading as jpeg, Ranges from >0 to 1.
+    */
+    public download(imageType: string = 'png', quality: number = 1): void {
+        let pack = {
+            name : this.name,
+            sets : {},
+            options: {},
+            frames : []
+        };
+
+        let sKeys = Object.keys(this.sets);
+        let oKeys = Object.keys(this.options);
+        let downloadPromise = new Promise ((success) => {
+            try {
+                for (let i = 0; i < sKeys.length; i++) {
+                    let channel: string = sKeys[i];
+                    let dt =  (this.sets as any)[channel];
+                    (pack.sets as any)[channel] = dt.getContext().canvas.toDataURL('image/' + imageType, quality);
+                }
+                for (let i = 0; i < oKeys.length; i++) {
+                    let opt: string = oKeys[i];
+                    (pack.options as any)[opt] = (this.options as any)[opt];
+                }
+                for (let i = 0; i < this.frames.length; i++) {
+                    let _f = this.frames[i];
+                    (pack.frames as Array<number>).push(_f.scale.x, _f.scale.y, _f.offset.x, _f.offset.y);
+                }
+            success(pack);
+            }catch (err) {
+                return;
+            }
+        });
+
+        downloadPromise.then((result) => {
+            let data = "data:text/json;charset=utf-8," + encodeURIComponent(JSON.stringify(result, null, 4));
+            let _a = document.createElement('a');
+            _a.setAttribute("href", data);
+            _a.setAttribute("download", this.name + "_texurePackage.json");
+            document.body.appendChild(_a);
+            _a.click();
+            _a.remove();
+        });
+    }
+
+    /**
+    * Public method to load a texturePacker JSON file.
+    * @param data of the JSON file in string format.
+    * @returns Promise<void>
+    */
+    public updateFromJSON(data: string): Promise<void> {
+        return new Promise ((resolve, reject) => {

```

Comment:
no need for a promse at all
code is synchronous

Reply:
Ok this is weird, all this was deleted on my last revision.  I must have done something wrong and reverted changes somehow...